### PR TITLE
feat(remap): enable remap to work with metric events

### LINKED ISSUE #5383: Implement remap::Object for LogEvent and MetricEvent
Ref #3550 

Implements [RFC 3550 - 2020-11-01](https://github.com/timberio/vector/blob/master/rfcs/2020-11-01-remap-metrics-support.md)

Currently `remap::Object` is implemented for `Event` and it assumes that that event is for a `LogEvent`. In order to allow Remap to work with metrics the implementation for `Event` needs te be separated out into separate implementations for `LogEvent` and `MetricEvent`.

The paths available to `MetricEvent` are limited with only the following fields:

  -  .name - The name of the metric.
  -  .namespace - The namespace for the metric.
  -  .timestamp - Timestamp of the metric.
  -  .kind - The kind of the metric - Incremental or Absolute
  -  .tags - A map containing the tags set for the metric.

### LINKED ISSUE #5387: Allow the Remap transform to work with metric events
Ref #3550 

Implements [RFC 3550 - 2020-11-01](https://github.com/timberio/vector/blob/master/rfcs/2020-11-01-remap-metrics-support.md)

Following from #5383, once `remap::Object` is implemented for MetricEvents, the remap transform can be updated to work with metrics too.

Note this will not do conversions between Metric and Log events, so a Metric coming in will still be a Metric going out, and the same for Logs.

### PR DESCRIPTION
**Pull request by FungusHumungus**
Closes #5383 
Closes #5387 

This enables Remap to work with metric events.

- Moves the `remap::Object` implementation to `LogEvent`.
- Implements `remap::Object` for `Metric`.
- Updates the Remap transform to work with both Log and Metric events.
- Updates the Remap conditions to work with both Log and Metric events.

It should be noted that this PR only allows manipulation of the metadata around Metrics. It is not possible to manipulate the metric data itself. I have raised #5521 to track manipulation of the actual data, should we decide to go down that route.

### CODE REVIEW

File: src/conditions/remap.rs [hunk 1]
```diff
@@ -57,7 +57,10 @@ impl Remap {
         // program wants to mutate its events.
         //
         // see: https://github.com/timberio/vector/issues/4744
-        Runtime::default().execute(&mut event.clone(), &self.program)
+        match event {
+            Event::Log(event) => Runtime::default().execute(&mut event.clone(), &self.program),
+            Event::Metric(event) => Runtime::default().execute(&mut event.clone(), &self.program),
+        }
```
Comment (JeanMertz) [inline src/conditions/remap.rs:63]: Not a big deal, but we can move some duplication out of this match by casing `event` to an `dyn Object` inside it, and then leave the rest of the line out of the match.

File: src/event/log_event.rs [hunk 2]
```diff
@@ -243,6 +243,84 @@ impl Serialize for LogEvent {
     }
 }
 
+impl remap::Object for LogEvent {
```
Comment (JeanMertz) [inline src/event/log_event.rs:246]: Just to check: I don't have to review this part right? It's only moved without any modifications?

File: src/event/metric.rs [hunk 3]
```diff
@@ -257,6 +261,21 @@ impl Metric {
...
     }
+
+    /// Sets or updates the string value of a tag
+    pub fn set_tag_value(&mut self, name: String, value: String) {
+        if self.tags.is_none() {
+            let map = BTreeMap::new();
+            self.tags = Some(map);
+        }
+
+        self.tags.as_mut().and_then(|tags| tags.insert(name, value));
```
Comment (JeanMertz) [inline src/event/metric.rs:272]: There's a `get_or_insert` on `Option` which should make this a bit easier to write.

File: src/event/metric.rs [hunk 4]
```diff
@@ -360,6 +379,112 @@ impl Display for Metric {
     }
 }
 
+impl Object for Metric {
+    fn insert(&mut self, path: &remap::Path, value: remap::Value) -> Result<(), String> {
+        if path.is_root() {
+            return Err("cannot set root path for a metric event".into());
```
Comment (JeanMertz) [inline src/event/metric.rs:385]: Given the amount of error-cases we'll need to handle here and below, I'd vote to move these into an enum (in Remap we use `thiserror` but we use `snafu` in Vector itself for that).

File: src/event/metric.rs [hunk 5]
```diff
@@ -360,6 +379,112 @@ impl Display for Metric {
...
+        }
+
+        match path.segments() {
+            [Segment::Field(tags), Segment::Field(field)] if tags.as_str() == "tags" => {
+                let value = value.as_bytes().ok_or("value must be a string")?;
```
Comment (JeanMertz) [inline src/event/metric.rs:390]: We have `try_bytes()` for this, which gives a more consistent error (e.g. "value must be string, but got ...")

(similar for all the other `ok_or`'s below)

File: src/event/metric.rs [hunk 6]
```diff
@@ -360,6 +379,112 @@ impl Display for Metric {
...
+                        Ok(())
+                    }
+                    "incremental" => {
+                        self.kind = MetricKind::Incremental;
+                        Ok(())
+                    }
+                    _ => Err("metric kind must be `absolute` or `incremental`".into()),
+                }
+            }
+            _ => Err("invalid path".into()),
```
Comment (JeanMertz) [inline src/event/metric.rs:426]: We probably want a bit more of a descriptive error message here.

One way to tackle this is to have a static list of whitelisted paths, and have this error report a message such as "got invalid path X, expected one of Y, Z"

File: src/event/metric.rs [hunk 7]
```diff
@@ -360,6 +379,112 @@ impl Display for Metric {
...
+            )),
+            [Segment::Field(tags), Segment::Field(field)] if tags.as_str() == "tags" => {
+                Ok(self.tag_value(field.as_str()).map(|value| value.into()))
+            }
+            _ => Err("invalid path".into()),
+        }
+    }
+
+    fn paths(&self) -> Result<Vec<remap::Path>, String> {
+        Ok(["name", "namespace", "timestamp", "tags"]
```
Comment (JeanMertz) [inline src/event/metric.rs:459]: This needs another iteration to also include the existing paths under `tags`, right?

Basically, we'd want:

```
.name
.namespace
.timestamp
.tags
.tags.tag1
.tags.tag2
```

File: src/event/metric.rs [hunk 8]
```diff
@@ -360,6 +379,112 @@ impl Display for Metric {
...
+            .iter()
+            .map(|path| Path::from_str(path).expect("invalid path"))
+            .collect())
+    }
+
+    fn remove(&mut self, path: &remap::Path, _compact: bool) -> Result<(), String> {
+        if path.is_root() {
+            return Err("cannot set root path for a metric event".into());
```
Comment (JeanMertz) [inline src/event/metric.rs:467]: s/set/remove (this would be its own enum variant if we move to a snafu-derived error object)

File: src/transforms/remap.rs [hunk 9]
```diff
@@ -72,8 +72,12 @@ impl FunctionTransform for Remap {
         emit!(RemapEventProcessed);
 
         let mut runtime = Runtime::default();
+        let result = match event {
+            Event::Log(ref mut event) => runtime.execute(event, &self.program),
+            Event::Metric(ref mut event) => runtime.execute(event, &self.program),
+        };
```
Comment (JeanMertz) [inline src/transforms/remap.rs:78]: Another minor nit that we should be able to keep `runtime.execute` out of the pattern matching.

File: tests/behavior/transforms/remap.toml [hunk 10]
```diff
@@ -1225,3 +1225,28 @@
...
     "a.equals" = "fbaro"
     "b.equals" = "fbarbar"
+
+[transforms.remap_metrics]
+  inputs = []
+  type = "remap"
+  source = """
+      .tags.host = "ook"
+      .tags.name = .name
+      .tags.namespace = .namespace
```
Comment (JeanMertz) [inline tests/behavior/transforms/remap.toml:1235]: This doesn't appear to be testing everything we want it to test, right?

That is, it's only testing inserting `tags` here, whereas we should also test`.namespace = xxx`, `.name = xxx` etc.

(I assume nesting `namespace` and `name` under `tags` was the oversight here?)

**Review by JeanMertz — CHANGES_REQUESTED** [review]
Sweet @FungusHumungus!

I love how relatively few changes we needed to make (actually none to TRL itself) to make it work with metrics 👌

I left a few comments that we'll likely want to address, but this is definitely close to being ready.

**Comment by JeanMertz** [comment]
@FungusHumungus reading this again, if I'm understanding this implementation correctly, this currently only allows modifying metric _metadata_, not the actual metric values, correct?

It's probably good to mention that in the PR description (and point to any tracking issues to iterate on this going forward).

File: src/conditions/remap.rs (hunk 1 above)
Comment (FungusHumungus) [reply src/conditions/remap.rs:63]: It would then need boxing etc.. I felt it wasn't worth the extra overhead to save the bit of code duplication.

File: src/event/log_event.rs (hunk 2 above)
Comment (FungusHumungus) [reply src/event/log_event.rs:246]: Yes correct.

File: src/conditions/remap.rs (hunk 1 above)
Comment (JeanMertz) [reply src/conditions/remap.rs:63]: Ah yeah, I see. I guess you could use generics (you can define a function within this function). But it's fine the way it is as well 😄

File: src/event/metric.rs (hunk 3 above)
Comment (FungusHumungus) [reply src/event/metric.rs:272]: Doh. Yes, I remember this came up in another review!

File: src/conditions/remap.rs (hunk 1 above)
Comment (FungusHumungus) [reply src/conditions/remap.rs:63]: That's a nice idea, but I can't get it to work..

```rust
        let execute = |object| Runtime::default().execute(object, &self.program);

        match event {
            Event::Log(event) => execute(&mut event.clone()),
            Event::Metric(event) => execute(&mut event.clone()),
        }
```
`execute` is monomorphized and so can't handle both paths. Casting to `dyn remap::Object` then falls down as it is unsized..

I could move execute into a top level function in the struct, but the benefits of doing so then rapidly diminish!

Comment (JeanMertz) [reply src/conditions/remap.rs:63]: Yeah, you'd have to use a function, as closure's can't take generic parameters. You can still define the function within this function, but I agree, the benefits aren't as high anymore.

File: src/event/metric.rs (hunk 7 above)
Comment (FungusHumungus) [reply src/event/metric.rs:459]: Oh, good catch. I misread how the function was working for `LogEvent`.

File: tests/behavior/transforms/remap.toml (hunk 10 above)
Comment (FungusHumungus) [reply tests/behavior/transforms/remap.toml:1235]: It seems that for metric events, the default contition only checks tags - it doesn't work with any other fields of the metric.

I could, of course use the Remap condition which can check all the relevant fields for the event - but it feels wrong to use the thing that you are testing to test the thing.

**Comment by FungusHumungus** [comment]
@JeanMertz Have made the requested changes.

I also added a `.type` field. From a [conversion an Discord](https://discordapp.com/channels/742820443487993987/746070604192415834/786988488948056084) it might be useful, and I can't see a downside.

File: lib/remap-lang/src/object.rs [hunk 11]
```diff
@@ -38,22 +68,22 @@ pub trait Object: std::fmt::Debug {
     /// When inserting into a coalesced path, the implementor is encouraged to
     /// insert into the right-most segment if none exists, but can return an
     /// error if needed.
-    fn insert(&mut self, path: &Path, value: Value) -> Result<(), String>;
+    fn insert(&mut self, path: &Path, value: Value) -> Result<(), Error>;
```
Comment (JeanMertz) [inline lib/remap-lang/src/object.rs:71]: I don't think exposing this type in this trait is the right approach to take.

I prefer we keep the error type of these functions free-form, such that the implementer can decide if they want to return a "structured error" or just some text.

What we _could_ do (and maybe should), is that we add an associated type to the trait:

```rust
 pub trait Object: std::fmt::Debug {
	type Err: std::fmt::Display;
}
```

This allows the implementer to use structured errors if they want to (such as is the case with the `impl Object for MetricEvent`), or use an unstructured `String`.

Basically, what I'm suggesting is we revert 733c44511509575895ec5a5094aa51c374d87a27, then move the structured errors into Vector itself, and have it set `type Err = Error`.

In remap we then just print the final string from the resulting error type we get back.

**Review by JeanMertz — CHANGES_REQUESTED** [review]

File: lib/remap-lang/src/object.rs (hunk 11 above)
Comment (JeanMertz) [reply lib/remap-lang/src/object.rs:71]: However, we might as well _not_ add the `Err` associated type, and just have them return a `String` (as it was before), as the callee can always call `to_string()` themselves before they pass their structured error to TRL.

**Comment by binarylogic** [comment]
Nice work!
